### PR TITLE
CO to change star rating with controller

### PR DESCRIPTION
**Pull request by ronso0**
This introduces controls
`[ChannelN],stars_up`
`[ChannelN],stars_down`
`[PreviewDeckN],stars_up`
`[PreviewDeckN],stars_down`
to allow changing a deck's star rating with controllers.
[lp:1814096](https://bugs.launchpad.net/mixxx/+bug/1814096)

Is this the correct approach?
No interference with mouse-over, stars are set in decks' visual and in the library immediately.

**ToDo**
- [x] make strings translatable in Controller Wizard
- [x] add the controls for the preview deck(s)
- [x] group controls properly in Controller Wizard > User Interface
- [x] clean up/improve starrating.cpp comments on drawing the polygons

I have a working implementation for the Reloop TerminalMix2/4 mapping which I'll clean up and commit here, as well:
- press & hold a deck's Load button
- turn the Trax knob to increase/decrease the respective deck's rating

### CODE REVIEW

**Comment by ronso0** [comment]
Where should I add the translate strings?
It seems https://github.com/mixxxdj/mixxx/blob/master/src/skin/tooltips.cpp is for skin tooltips only?

File: src/library/starrating.cpp [hunk 1]
```diff
@@ -20,14 +20,26 @@
...
         : m_myStarCount(starCount),
           m_myMaxStarCount(maxStarCount) {
+    // 1st star cusp at 0° of the unit circle whose center is shifted to adapt the 0,0-based paint area
     m_starPolygon << QPointF(1.0, 0.5);
-    for (int i = 1; i < 5; ++i)
+    for (int i = 1; i < 5; ++i) {
+        // add QPointF 2-5 to polygon point array, equally distributed on a circumference.
+        // To create a star (not a pentagon) we need to connect every second point.
+        //
+        // understand those equations
```
Comment (ronso0) [inline src/library/starrating.cpp]: ..which I did after some trigonometry coaching. I'm still puzzled why the stars are pointing right, though. Seems like `x` and `y` are swapped.. When I use the debug output to draw a chart in libreoffice calc the star points upwards.

File: src/library/starrating.cpp [hunk 2]
```diff
@@ -20,14 +20,26 @@
 #include "library/starrating.h"
 #include "util/math.h"
 
+// Magic number? Explain what this factor affects and how
```
Comment (ronso0) [inline src/library/starrating.cpp]: AFAICT this closely follows the [QT example](https://doc.qt.io/qt-5.6/qtwidgets-itemviews-stardelegate-example.html), which doesn't provide a low-level explanation either, so I'd like to add some explanations.

**Comment by foss-** [comment]
Love this idea and it reminds me that one of my top feature wish list items still is for mixxx to write back star ratings to the files metadata info. There was an old PR for that but I think it got closed during a cleanup.

**Comment by ronso0** [comment]
I guess you refer to https://bugs.launchpad.net/mixxx/+bug/1799441

File: src/controllers/controlpickermenu.cpp [hunk 3]
```diff
@@ -802,6 +802,20 @@ ControlPickerMenu::ControlPickerMenu(QWidget* pParent)
     addDeckControl("waveform_zoom", tr("Waveform Zoom"), tr("Waveform zoom"), guiMenu);
     addDeckControl("waveform_zoom_down", tr("Waveform Zoom In"), tr("Zoom waveform in"), guiMenu);
     addDeckControl("waveform_zoom_up", tr("Waveform Zoom Out"), tr("Zoom waveform out"), guiMenu);
+
+    // Controls to change a deck's star rating
+    QString starsUpTitle = tr("Star Rating Up");
+    QString starsUpDescription = tr("Increase the track rating by one star");
```
Comment (daschuer) [inline src/controllers/controlpickermenu.cpp]: Just out of curious: 
For my German ears "Upgrades the track rating .. " sounds better. I would use "Increase" for floating point values. 
What do the native speaker say?

File: src/widget/wstarrating.cpp [hunk 4]
```diff
@@ -87,6 +92,32 @@ void WStarRating::mouseMoveEvent(QMouseEvent *event) {
     }
 }
 
+void WStarRating::slotStarsUp(double v) {
+    if (!m_pCurrentTrack) {
+        return;
+    }
+    if (v > 0 && m_starRating.starCount() < m_starRating.maxStarCount()) {
+        int star = m_starRating.starCount() + 1;
+        qDebug() << "   stars " << m_starRating.starCount() << " > " << star;
```
Comment (daschuer) [inline src/widget/wstarrating.cpp]: why we have extra space around stars? qDebug() already places spaces around << added strings.

File: src/library/starrating.cpp [hunk 5]
```diff
@@ -20,14 +20,26 @@
...
+    // 1st star cusp at 0° of the unit circle whose center is shifted to adapt the 0,0-based paint area
     m_starPolygon << QPointF(1.0, 0.5);
-    for (int i = 1; i < 5; ++i)
+    for (int i = 1; i < 5; ++i) {
+        // add QPointF 2-5 to polygon point array, equally distributed on a circumference.
+        // To create a star (not a pentagon) we need to connect every second point.
+        //
+        // understand those equations
         m_starPolygon << QPointF(0.5 + 0.5 * cos(0.8 * i * 3.14), 0.5 + 0.5 * sin(0.8 * i * 3.14));
+        float QPointFx = 0.5 + 0.5 * cos(0.8 * i * 3.14);
```
Comment (daschuer) [inline src/library/starrating.cpp]: can this also be commented? Q is reserved for Qt functions and we start variables with a lower case by the way.

File: src/widget/wstarrating.cpp (hunk 4 above)
Comment (ronso0) [reply src/widget/wstarrating.cpp]: This is left-over debug output I will remove

File: src/library/starrating.cpp (hunk 5 above)
Comment (ronso0) [reply src/library/starrating.cpp]: debug output I will remove

File: src/widget/wstarrating.cpp (hunk 4 above)
Comment (ronso0) [reply src/widget/wstarrating.cpp]: white the leading whitespace the output is easier to spot in the log stream

**Comment by ronso0** [comment]
okay, this how it looks like right now in the controller Lerning Wizard, listed are all available decks & preview decks:
star-co-midi-wizard

**Comment by ronso0** [comment]
how do I make the strings translatable?
I could reverse-engineer this as well, but it would be nice if someone could just show the shortcut.

**Comment by daschuer** [comment]
The tr() function is sufficient.
Once this is merged, we will upload the strings to transiflex, and get hopefully the translations back if someone takes care. 


Is this ready for merge now?

**Comment by ronso0** [comment]
> Is this ready for merge now?

Jup!

I'll polish the Reloop implementaion in another PR.

File: src/controllers/controlpickermenu.cpp [hunk 6]
```diff
@@ -153,6 +153,7 @@ ControlPickerMenu::ControlPickerMenu(QWidget* pParent)
     QMenu* eqMenu = addSubmenu(tr("Equalizers"));
     const int kNumEqRacks = 1;
     const int iNumDecks = ControlObject::get(ConfigKey("[Master]", "num_decks"));
+    const int iNumPreviewDecks = ControlObject::get(ConfigKey("[Master]", "num_preview_decks"));
```
Comment (ronso0) [inline src/controllers/controlpickermenu.cpp]: will remove this. not necesary anymore since 'addDeckAndPreviewDeckControl()' is used for star COs below

**Comment by ronso0** [comment]
Done.

**Comment by daschuer** [comment]
LGTM Thank you.
Please take also care to update this: https://www.mixxx.org/wiki/doku.php/mixxxcontrols

**Comment by ronso0** [comment]
> Please take also care to update this: https://www.mixxx.org/wiki/doku.php/mixxxcontrols

Done.

**Pull request merged**